examples/with-apollo-auth/lib/apollo.js

```diff
@@ -0,0 +1,195 @@
+          return parseCookies().token
+        }
+      })
+    }, [])
+    return (
+      <ApolloProvider client={client}>
+        <PageComponent {...pageProps} />
+      </ApolloProvider>
+    )
+  }
+
+  if (process.env.NODE_ENV !== 'production') {
+    // Find correct display name
+    const displayName =
+      PageComponent.displayName || PageComponent.name || 'Component'
+
+    // Warn if old way of installing apollo is used
+    if (displayName === 'App') {
+      console.warn('This withApollo HOC only works with PageComponents.')
+    }
+
+    // Set correct display name for devtools
+    WithApollo.displayName = `withApollo(${displayName})`
+
+    // Add some prop types
+    WithApollo.propTypes = {
+      // Used for getDataFromTree rendering
+      apolloClient: PropTypes.object,
+      // Used for client/server rendering
+      apolloState: PropTypes.object
+    }
+  }
+
+  WithApollo.getInitialProps = async ctx => {
+    const { AppTree, req, res } = ctx
+
+    // Run all GraphQL queries in the component tree
+    // and extract the resulting data
+    const apolloClient = (ctx.apolloClient = initApolloClient(
+      {},
+      {
+        getToken: () => parseCookies(req).token
+      }
+    ))
+
+    const pageProps = PageComponent.getInitialProps
+      ? await PageComponent.getInitialProps(ctx)
+      : {}
+
+    if (res && res.finished) {
+      // When redirecting, the response is finished.
+      // No point in continuing to render
+      return {}
+    }
+
+    // Get apolloState on the server (needed for ssr)
+    if (typeof window === 'undefined') {
+      try {
+        // Run all GraphQL queries
+        const { getDataFromTree } = await import('@apollo/react-ssr')
+        await getDataFromTree(
+          <AppTree
+            pageProps={{
+              ...pageProps,
+              apolloClient
+            }}
+          />
+        )
+      } catch (error) {
+        // Prevent Apollo Client GraphQL errors from crashing SSR.
+        // Handle them in components via the data.error prop:
+        // https://www.apollographql.com/docs/react/api/react-apollo.html#graphql-query-data-error
+        console.error('Error while running `getDataFromTree`', error)
+      }
+
+      // getDataFromTree does not call componentWillUnmount
+      // head side effect therefore need to be cleared manually
+      Head.rewind()
+    }
+
+    // Extract query data from the Apollo store
+    const apolloState = apolloClient.cache.extract()
+
+    return {
+      ...pageProps,
+      apolloState
+    }
+  }
+
+  return WithApollo
+}
+
+let apolloClient = null
+
+/**
+ * Always creates a new apollo client on the server
+ * Creates or reuses apollo client in the browser.
+ */
+function initApolloClient (...args) {
+  // Make sure to create a new client for every server-side request so that data
+  // isn't shared between connections (which would be bad)
+  if (typeof window === 'undefined') {
+    return createApolloClient(...args)
```

Comment:
Any way to make Typescript accept the `...args` ? It seems to get cranky with the spread operator...